quantecon/lss.py

```diff
@@ -406,3 +416,61 @@ def impulse_response(self, j=5):
             Apower = np.dot(Apower, A)
 
         return xcoef, ycoef
+
+    def __partition(self):
+        r"""
+        Reorder the states by shifting the constant terms to the
+        top of the state vector. Then partition the linear state
+        space model following Appendix C in RMT Ch2 such that the
+        A22 matrix associated with non-constant states have eigenvalues
+        all strictly smaller than 1.
+
+        .. math::
+            \left[\begin{array}{c}
+            const\\
+            x_{2,t+1}
+            \end{array}\right]=\left[\begin{array}{cc}
+            I & 0\\
+            A_{21} & A_{22}
+            \end{array}\right]\left[\begin{array}{c}
+            1\\
+            x_{2,t}
+            \end{array}\right]+\left[\begin{array}{c}
+            0\\
+            C_{2}
+            \end{array}\right]w_{t+1}
+
+        Returns
+        -------
+        A22 : array_like(float)
+            Lower right part of the reordered matrix A.
+        A21 : array_like(float)
+            Lower left part of the reordered matrix A.
+        """
+        A, C = self.A, self.C
+        n = self.n
+
+        sorted_idx = []
+        A_diag = np.diag(A)
+        num_const = 0
+        for idx in range(n):
+            if (A_diag[idx] == 1) and (C[idx, :] == 0).all():
```

Comment:
I agree that this part is actually tricky. But I guess we need to do this inside the function so that we don't complicate the input format, with the same logic you mentioned above.

Essentially we want to detect those states associated with unity eigenvalues (we do not allow for eigenvalues strictly greater than 1). In addition to the current two conditions for finding the targeted states, I guess we can check `np.linalg.norm(row_vec)==1` or we can compute the eigenvalues directly.

Reply:
The second example I gave has eigenvalues 1 and 0, and none of the rows satisfy `np.linalg.norm(row_vec)==1`. Besides, a row may satisfy `np.linalg.norm(row_vec)==1`, but not be a vector containing one 1 and 0s otherwise. 

I think the following procedure might be good enough, even if it doesn't cover all possible cases:
1. Check whether the ith row is a vector with a 1 in the ith entry, and 0s otherwise. Store these rows as "constant" rows.
2. Re-order the matrices based on 1.
3. Verify that A_22 is stable by computing the modulus of eigenvalues. If the check fails, raise an error.

What do you think?

Reply:
Isn't step 1 equivalent to `(A_diag[idx] == 1) and np.linalg.norm(row_vec)==1`? Sorry that I might not clear in my previous reply, I meant to add `np.linalg.norm(row_vec)==1` "additionally". And `(C[idx, :] == 0).all()` is also a necessary condition to check for a constant state.

Reply:
Yes, this looks good to me. I thought you meant only checking `np.linalg.norm(row_vec)==1`.